crates/core_simd/src/cast.rs

```diff
@@ -0,0 +1,129 @@
+use crate::simd::{intrinsics, LaneCount, Simd, SimdElement, SupportedLaneCount};
+
+/// Supporting trait for `Simd::cast`.  Typically doesn't need to be used directly.
+pub trait SimdCast<Target: SimdElement>: SimdElement {
+    #[doc(hidden)]
+    fn cast<const LANES: usize>(x: Simd<Self, LANES>) -> Simd<Target, LANES>
+    where
+        LaneCount<LANES>: SupportedLaneCount;
+}
+
+macro_rules! into_number {
+    { $from:ty, $to:ty } => {
+        impl SimdCast<$to> for $from {
+            fn cast<const LANES: usize>(x: Simd<Self, LANES>) -> Simd<$to, LANES>
+            where
+                LaneCount<LANES>: SupportedLaneCount,
+            {
+                // Safety: simd_as can handle numeric conversions
+                unsafe { intrinsics::simd_as(x) }
+            }
+        }
+    };
+    { $($type:ty),* } => {
+        $(
+        into_number! { $type, i8 }
+        into_number! { $type, i16 }
+        into_number! { $type, i32 }
+        into_number! { $type, i64 }
+        into_number! { $type, isize }
+
+        into_number! { $type, u8 }
+        into_number! { $type, u16 }
+        into_number! { $type, u32 }
+        into_number! { $type, u64 }
+        into_number! { $type, usize }
+
+        into_number! { $type, f32 }
+        into_number! { $type, f64 }
+        )*
+    }
+}
+
+into_number! { i8, i16, i32, i64, isize, u8, u16, u32, u64, usize, f32, f64 }
+
+macro_rules! into_pointer {
+    { $($type:ty),* } => {
+        $(
+        impl<T> SimdCast<$type> for *const T {
+            fn cast<const LANES: usize>(x: Simd<Self, LANES>) -> Simd<$type, LANES>
+            where
+                LaneCount<LANES>: SupportedLaneCount,
+            {
+                // Safety: transmuting isize to pointers is safe
+                let x: Simd<isize, LANES> = unsafe { core::mem::transmute_copy(&x) };
```

Comment:
I would consider that questionably safe, pointer to integer transmutes are almost certainly UB:
https://github.com/rust-lang/unsafe-code-guidelines/issues/286

Reply:
also, the safety comment is the wrong way around, you're transmuting pointers to integers here, which is probably UB.

Reply:
It's no longer almost certainly UB, but it does lose provenance (and `as`/casting should not): https://doc.rust-lang.org/nightly/src/core/ptr/const_ptr.rs.html#183-185

Reply:
i meant its UB to use the result by converting it back to a pointer and expecting it to be valid.

Reply:
I see. That is still UB (for a while, the transmute itself was probably UB too).

Reply:
My understanding is that once you convert a pointer to an integer, it loses provenance no matter what, and you need to use `with_addr` to inherit an existing pointer's provenance.  So this particular line (converting a pointer to an integer) should be fine, I think.

Reply:
that's actually incorrect -- on most platforms using `as` to cast a pointer to an integer exposes the pointer such that `as` casts from integers back to pointers using that address produces the original pointer including provenance. transmute doesn't do that.

because these functions are named `cast` and intended to function like `as`, they need to have the same pointer exposing behavior as `as`. you don't get that with `transmute`

Reply:
I realized my confusing when reviewing the docs for `addr`:
> This is equivalent to `self as usize`, which semantically discards provenance and address-space information. To properly restore that information, use [with_addr](https://doc.rust-lang.org/std/primitive.pointer.html#method.with_addr) or [map_addr](https://doc.rust-lang.org/std/primitive.pointer.html#method.map_addr).

The bit about `self as usize` is incorrect, then?

Reply:
> I realized my confusing when reviewing the docs for `addr`:
> 
> > This is equivalent to `self as usize`, which semantically discards provenance and address-space information. To properly restore that information, use [with_addr](https://doc.rust-lang.org/std/primitive.pointer.html#method.with_addr) or [map_addr](https://doc.rust-lang.org/std/primitive.pointer.html#method.map_addr).
> 
> The bit about `self as usize` is incorrect, then?

yes, when referring to current `ptr as usize` semantics...that said, iirc they want to move to `as` being equivalent to `addr` rather than `expose_addr` at some point (requires edition boundary).